Refactor syntax unicode answer

livecodesebastien opened the pull request.

livecodepanos committed:
Some more updates on methods at 'Extras' tab

livecodepanos committed:
Updates after Seb's feedback

livecodepanos committed:
Compiles on mobile, but a 'registration error' window pops up in the beginning

livecodepanos committed:
Updated some signatures so as to compile on Linux

livecodepanos committed:
Still does not compile on Linux

livecodepanos committed:
Updated MCS_resolvepath() signature so as to compile on Linux

livecodepanos committed:
Now it compiles on Linux

livecodepanos committed:
Updates on lnxans.cpp

livecodepanos committed:
Fixed bug in Linux (the one that no title appeared in the window generated by the test script

livecodesebastien committed:
EP functions from refactor-syntax_unicode_EPConstructors2

livecodesebastien committed:
EP functions from refactor-syntax_unicode_GetEP1

livecodesebastien committed:
EP functions from refactor-syntax_unicode_GetEP2

livecodesebastien committed:
EP functions from refactor-syntax_unicode_objecteval

livecodesebastien committed:
Deleted garbage character

livecodepanos committed:
Added ctxt . CopyAsArrayRef(..)

livecodesebastien committed:
Merge pull request #281 from runrevpanos/refactor-syntax_unicode_ctxt_functions

Added ctxt . CopyAsArrayRef(..)

livecodesebastien committed:
MCExecContext::Clear() added

livecodepanos committed:
Added ctxt.CopyAsNameRef() and ctxt.GetReal8() methods

livecodesebastien committed:
Merge pull request #282 from runrevpanos/refactor-syntax_unicode_ctxt_functions

Added ctxt.CopyAsNameRef() and ctxt.GetReal8() methods

livecodesebastien committed:
MCExecContext functions added
Clear()
IsEmpty()
IsArray()
IsString()
IsArray()
IsNumber()

livecodesebastien committed:
Answer and Aks functions refactored
Only Linux is implemented for now

livecodesebastien committed:
Answer and Aks file/folder
Mac functions implemented

livecodesebastien committed:
Answer and Ask file/folder
Windows functions implemented
Complete suppression of the ExecContext formerly used to pass the result

livecodesebastien committed:
Answer and Ask functions
Mac implementation: signatures updated to EP-less MCA_* functions

livecodesebastien committed:
Linux file chooser improvement
'answer file <windowtitle> with <filename>' now open <filename> at the path specified even if it's a folder

livecodesebastien committed:
Merge branch 'refactor-syntax_unicode' into refactor-syntax_unicode_answer

Conflicts:
	engine/src/dskw32.cpp
	engine/src/exec.h

Fixed error in Linux - ask file always returned false

runrevmark commented on engine/src/exec-dialog.cpp:
```diff
@@ -136,64 +136,39 @@ void MCDialogExecAnswerColor(MCExecContext &ctxt, MCColor *p_initial_color, MCSt
-	else
-		return ep.copyasstringref(r_value);
-}
+///* WRAPPER */
```
Remove this code if its obsolete.

runrevmark commented on engine/src/exec-dialog.cpp:
```diff
@@ -314,26 +294,26 @@ void MCDialogExecAnswerFileWithTypes(MCExecContext &ctxt, bool p_plural, MCStrin
-	else
-		return ep.copyasstringref(r_value);	
-}
+//static bool MCA_folder(bool p_plural, MCStringRef p_prompt, MCStringRef p_initial, MCStringRef p_title, bool p_sheet, MCStringRef& r_value, MCStringRef& r_result)
```
Remove this code if its obsolete.

runrevmark commented on engine/src/exec-dialog.cpp:
```diff
@@ -564,25 +547,25 @@ void MCDialogExecAskPassword(MCExecContext& ctxt, bool p_clear, MCStringRef p_pr
-	t_options = 0;
-	if (p_as_sheet)
-		t_options |= MCA_OPTION_SHEET;
+//static bool MCA_ask_file(MCStringRef p_prompt, MCStringRef p_initial, MCStringRef p_filter, MCStringRef p_title, bool p_as_sheet, bool& r_cancelled, MCStringRef& r_filename)
```
Remove this code if its obsolete.

runrevmark commented on engine/src/exec-dialog.cpp:
```diff
@@ -628,41 +620,41 @@ void MCDialogExecAskFileWithFilter(MCExecContext& ctxt, MCStringRef p_prompt, MC
-	t_options = 0;
-	if (p_as_sheet)
-		t_options |= MCA_OPTION_SHEET;
+//static bool MCA_ask_file_with_types(MCStringRef p_title, MCStringRef p_prompt, MCStringRef *p_types, uindex_t p_type_count, MCStringRef p_initial, bool p_as_sheet, bool& r_cancelled, MCStringRef& r_filename, MCStringRef& r_chosen_type)
```
Remove this code if its obsolete.

runrevmark commented on engine/src/lnxans.cpp:
```diff
@@ -311,40 +312,38 @@ static gboolean gtk_idle_callback (gpointer data)
 	if (gtk_dialog_run(GTK_DIALOG (dialog)) == GTK_RESPONSE_ACCEPT)
 	{
-
+        MCAutoStringRef t_filename;
```
Since we're building a list here, I suggest using MCListRef and MCListCopyAsStringRef logic.

runrevmark commented on engine/src/lnxans.cpp:
```diff
@@ -465,59 +475,66 @@ void set_initial_file ( GtkWidget *dialog, const char * p_initial, char * p_last
-		}
-		delete t_resolved_path;
+            MCRemoteFileDialog(p_title, p_prompt, t_rtypes, t_count, NULL, *t_resolved_path, false, t_plural, r_result);
+//			MCCStringArrayFree(t_rtypes, p_type_count * 2);
```
If types_to_remote_types is returning a MCStringRef*, then you'll need to release each element of the array (assuing the array owns them).

runrevmark commented on engine/src/lnxans.cpp:
```diff
@@ -560,33 +576,35 @@ int MCA_file_with_types(MCExecPoint& ep, const char *p_title, const char *p_prom
-		}
-		delete t_resolved_path;
+            MCRemoteFileDialog(p_title, p_prompt, t_rtypes, t_count, NULL, *t_resolved_path, true, t_plural, r_result);
+            MCMemoryDeleteArray(t_rtypes); // memory leak?
```
Same comment as above - MCCStringArrayFree would delete each element then the array itself.

runrevmark commented on engine/src/osxans.mm:
```diff
@@ -498,17 +509,17 @@ - (void) panelDidEnd:(NSSavePanel *)savePanel returnCode:(int)returnCode conextI
 	t_initial_folder = nil;
 	if (p_initial_folder != nil)
-		t_initial_folder = [NSString stringWithCString: p_initial_folder encoding: NSMacOSRomanStringEncoding];
+        t_initial_folder = [NSString stringWithCString: MCStringGetCString(p_initial_folder) encoding: NSMacOSRomanStringEncoding];
```
You can use stringWithMCStringRef: for these now :)

livecodesebastien committed:
Update after feedback:
* Array deallocation handled properly
* Removed commented out code
* Creation of filename list done using MCListRef now

livecodesebastien committed:
Answer and Ask:
Fixed result on Cancel

livecodesebastien committed:
Answer ans Ask

Fixed bug on Windows - filters were not working
Fixed error on return on Windows - was always returning Cancel.

livecodesebastien committed:
Answer and Ask
Fixed result return value on Linux

livecodesebastien committed:
Merge branch 'refactor-syntax_unicode' into refactor-syntax_unicode_answer

livecodesebastien committed:
Reverted unnecessary changes in kernel.vcproj, dskw32.cpp and exec.h
Added forgotten SetTheResultToValue in MCDialogExecCustomAskDialog

livecodesebastien committed:
Merge branch 'refactor-syntax_unicode_extras5' into refactor-syntax_unicode_answer

Conflicts:
	engine/src/lnxans.cpp
	engine/src/mode_development.cpp
	engine/src/mode_installer.cpp
	engine/src/mode_server.cpp
	engine/src/mode_standalone.cpp
	engine/src/osxans.mm
	engine/src/osxansold.cpp
	engine/src/w32ans.cpp

Fixed bug on MCA_color which made the color picker useless - always returning 0,0,0
Reverted changes to mbldialog.cpp as the file is utterly commented out
